### DISCUSSION TITLE
[API Proposal]: The request for using a mixed mode interface during instance creation.

### DISCUSSION THREAD

**Discussion by bunnyi116** [dotnet/csharplang#7319] [Language Ideas]
```C#
var testClass = new TestClass("Username", "Password")
{
    Property1 = 1,
    Property2 = 2,
    // ...
};

// ...

class TestClass
{
    public required string Username { get; init; }
    public required string Password { get; init; }
    public required int Property1 { get; init; }
    public required long Property2 { get; init; }
    // ...... 
    
 	public TestClass() { }
    
    [SetsRequiredMembers]
    public TestClass(string username, string password, int property3, long property4)
    {
        Username = username;
        Password = password;
        Property1 = property3;
        Property2 = property4;
    }
}
```

I hope "required" can be mixed used in the constructor. As shown below, `"Username", "Password"` are initialized within parentheses (), while the remaining parts are initialized within {} curly brackets.

```
var testClass = new TestClass("Username", "Password")	
{
    Property1 = 1,
    Property2 = 2,
    // ...
};
```

**Comment by huoyaoyuan** [comment]
This is a feature request of C# language and should go to csharplang repo.

This was proposed in the original proposal, and the language team feel it doesn't fit into the initial implementation.

**Comment by TahirAhmadov** [comment]
How would this work? Does the compiler emit multiple constructors? What if the constructor has any other logic other than assigning to properties? C# does not (almost always) modify the signature of a member basing on its body, let alone basing on other members in the same class.

**Comment by bunnyi116** [comment]
This should be a syntactic sugar, where the constructor can be passed either by signature or in the order of the constructor’s signature. The remaining parameters can be passed using {}. During compilation, it checks if all of them are passed. If not, it throws an error. If all are passed, it transpiles into a normal constructor invocation at compile time.

**Reply by HaloFour** [reply]
The compiler shouldn't be trying to make any assumptions regarding whether or not properties map to constructor arguments, especially given that the naming conventions for the two are different.  This concept was explicitly rejected during the design of records.
